### PR TITLE
Allow Translations in the API

### PR DESCRIPTION
**Pull request by rjmackay**
This pull request makes the following changes:
- Move Usecase Traits into Usecase\Concerns namespace
- Remove unnecessary ...Data classes
- Add translations table
- Add Concerns\SavesTranslations Trait which saves translations via Repo
- Save translations on Create and Update
- Add Translations Repo
- Set translatable fields on Tags
- [x] Retrieve translations in formatter

Test checklist:
- [ ] bin/behat

Refs #1933

Ping @ushahidi/platform

### CODE REVIEW

**Comment by rjmackay** [comment]
@willdoran not finished yet, but could use a review anyway

**Comment by rjmackay** [comment]
Got a little carried away, so this is now mostly done... but I should sleep now!

File: migrations/20170720002747_add_translations_table.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    /**
+     * Change method
+     */
+    public function change()
+    {
+        $this->table('translations')
+            ->addColumn('resource', 'string', ['limit' => 50, 'null' => false])
+            ->addColumn('resource_id', 'integer', ['null' => false])
+            ->addColumn('property', 'string', ['limit' => 50, 'null' => false])
+            ->addColumn('source', 'string')
```
Comment (willdoran) [inline migrations/20170720002747_add_translations_table.php]: At some point we're going to want to mine translations, what is source in reference to?

File: src/Core/Usecase/Concerns/FilterRecords.php [hunk 2]
```diff
@@ -17,7 +17,7 @@
  * @license    https://www.gnu.org/licenses/agpl-3.0.html GNU Affero General Public License Version 3 (AGPL3)
  */
 
-namespace Ushahidi\Core\Traits;
+namespace Ushahidi\Core\Usecase\Concerns;
```
Comment (willdoran) [inline src/Core/Usecase/Concerns/FilterRecords.php]: 👍

File: src/Core/Usecase/Concerns/SavesTranslations.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	use TranslationRepositoryTrait;
+
+	protected function saveTranslations(Entity $entity) {
+		if (($translations = $this->getPayload('translations', false)) &&
+		    ($properties = $entity->getTranslatable())
+		) {
+			foreach ($translations as $locale => $values) {
+				foreach ($properties as $property) {
+					if (isset($values[$property])) {
+						$this->translationRepo->saveTranslation($entity->getResource(), $entity->getId(), $property, $values[$property], $entity->$property, $locale);
```
Comment (willdoran) [inline src/Core/Usecase/Concerns/SavesTranslations.php]: Will these be individual sql queries? Is there any benefit in trying to batch them?

File: tests/integration/tags.feature [hunk 4]
```diff
@@ -287,3 +287,30 @@ Feature: Testing the Tags API
...
+                "tag":"Boxes",
+                "slug":"boxes",
+                "description":"Is this a box? Awesome",
+                "type":"category",
+                "priority":1,
+                "color":"00ff00",
+                "role":[],
+                "translations":{
+                    "ar": {
+                        "tag": "Boxes in arabic"
```
Comment (willdoran) [inline tests/integration/tags.feature]: Would be worth using utf chars to make sure we don't get any weirdness when saving.

File: migrations/20170720002747_add_translations_table.php (hunk 1 above)
Comment (rjmackay) [inline migrations/20170720002747_add_translations_table.php]: Source is the source text. We don't need it now, but just for future reference and possibly flagging when the source changes

File: src/Core/Usecase/Concerns/SavesTranslations.php (hunk 3 above)
Comment (rjmackay) [inline src/Core/Usecase/Concerns/SavesTranslations.php]: I think I might try to batch it. I was trying to keep the repo dumb...

File: tests/integration/tags.feature (hunk 4 above)
Comment (rjmackay) [inline tests/integration/tags.feature]: Yes!

File: src/Core/Usecase/Concerns/SavesTranslations.php (hunk 3 above)
Comment (rjmackay) [inline src/Core/Usecase/Concerns/SavesTranslations.php]: Ok batching is hard while doing an UPSERT... Gonna ignore it for now

**Comment by willdoran** [comment]
@rjmackay is this ready to review or did you want to add more tests first?

**Comment by rjmackay** [comment]
@willdoran I think you reviewed everything here already, its just been rebased since.
The biggest gap is that we don't validate values submitted for translations ie. checking length of title/description etc... Do you think we could just land anyway and address that later? or should I park till thats done?

**Comment by willdoran** [comment]
@rjmackay Yeah I think landing is reasonable.

**Comment by caratzky** [comment]
@rjmackay  do you need translations for your project? There are a lot of Utopian Contributors who would gladly do it for you, because we are well-rewarded for the job done. I, for example, can do translations in my native language, Tagalog (tl), Filipino (fil) and Cebuano (ceb) which are all listed in ISO 639-1. I'm hoping to hear from you soon. Thank you.

Best Regards,
Caratzky

**Comment by tuxpiper** [comment]
@caratzky  we gladly accept translation submissions from community members. Head over to https://www.transifex.com/ushahidi/ushahidi-v3 if you are inclined to contribute. Thank you!

**Comment by rowasc** [comment]
https://github.com/ushahidi/platform/issues/1933 linked here. Closing since we will need to work on it to  add the client , migrated this work to lumen so we can progress it

**Pull request closed without merging**